test/e2e/singlecluster/pod_test.go

```diff
@@ -0,0 +1,145 @@
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/client-go/discovery"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/controller/jobs/pod"
+	"sigs.k8s.io/kueue/pkg/util/kubeversion"
+	"sigs.k8s.io/kueue/pkg/util/testing"
+	podtesting "sigs.k8s.io/kueue/pkg/util/testingjobs/pod"
+	"sigs.k8s.io/kueue/test/util"
+)
+
+// +kubebuilder:docs-gen:collapse=Imports
+
+var _ = ginkgo.Describe("Pod groups", func() {
+	var (
+		ns         *corev1.Namespace
+		onDemandRF *kueue.ResourceFlavor
+	)
+
+	ginkgo.BeforeEach(func() {
+		if kubeVersion().LessThan(kubeversion.KubeVersion1_27) {
+			ginkgo.Skip("Unsupported in versions older than 1.27")
+		}
+		ns = &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "pod-e2e-",
+			},
+		}
+		gomega.Expect(k8sClient.Create(ctx, ns)).To(gomega.Succeed())
+		onDemandRF = testing.MakeResourceFlavor("on-demand").Label("instance-type", "on-demand").Obj()
+		gomega.Expect(k8sClient.Create(ctx, onDemandRF)).To(gomega.Succeed())
+	})
+	ginkgo.AfterEach(func() {
+		gomega.Expect(util.DeleteNamespace(ctx, k8sClient, ns)).To(gomega.Succeed())
+		util.ExpectResourceFlavorToBeDeleted(ctx, k8sClient, onDemandRF, true)
+	})
+
+	ginkgo.When("Single CQ", func() {
+		var (
+			cq *kueue.ClusterQueue
+			lq *kueue.LocalQueue
+		)
+
+		ginkgo.BeforeEach(func() {
+			cq = testing.MakeClusterQueue("cq").
+				ResourceGroup(
+					*testing.MakeFlavorQuotas("on-demand").Resource(corev1.ResourceCPU, "5").Obj(),
+				).
+				Obj()
+			gomega.Expect(k8sClient.Create(ctx, cq)).To(gomega.Succeed())
+			lq = testing.MakeLocalQueue("queue", ns.Name).ClusterQueue(cq.Name).Obj()
+			gomega.Expect(k8sClient.Create(ctx, lq)).To(gomega.Succeed())
+		})
+		ginkgo.AfterEach(func() {
+			util.ExpectClusterQueueToBeDeleted(ctx, k8sClient, cq, true)
+		})
+
+		ginkgo.It("should admit group that fits", func() {
+			g1 := podtesting.MakePod("g1", ns.Name).
+				Image("gcr.io/k8s-staging-perf-tests/sleep:v0.0.3", []string{"1s"}).
+				Queue(lq.Name).
+				Request(corev1.ResourceCPU, "1").
+				MakeGroup(2)
+			g1Key := client.ObjectKey{Namespace: ns.Name, Name: "g1"}
+			for _, p := range g1 {
+				gomega.Expect(k8sClient.Create(ctx, p)).To(gomega.Succeed())
+				gomega.Expect(p.Spec.SchedulingGates).
+					To(gomega.ContainElement(corev1.PodSchedulingGate{
+						Name: pod.SchedulingGateName}))
+			}
+			ginkgo.By("Starting admission", func() {
+				util.UnholdQueue(ctx, k8sClient, cq)
+
+				// Verify that the Pods start with the appropriate selector.
+				gomega.Eventually(func(g gomega.Gomega) {
+					for _, origPod := range g1 {
+						var p corev1.Pod
+						gomega.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(origPod), &p)).To(gomega.Succeed())
+						g.Expect(p.Spec.SchedulingGates).To(gomega.BeEmpty())
+						g.Expect(p.Spec.NodeSelector).To(gomega.Equal(map[string]string{
```

Comment:
Question: the pods execute for 1s only, will this check pass if executed after 1s?

Reply:
This is a good point.
This check would pass since the kueue-manager doesn't restore nodeSelectors: https://github.com/kubernetes-sigs/kueue/blob/4bc332a75fe221e6dc92db269ae3832da884527a/pkg/controller/jobs/pod/pod_controller.go#L179-L183

However, if the pods finished before Line104 is performed, this check wouldn't pass since the manager will remove completed pods: https://github.com/kubernetes-sigs/kueue/blob/4bc332a75fe221e6dc92db269ae3832da884527a/pkg/controller/jobs/pod/pod_controller.go#L329-L331

@alculquicondor So, I guess that we should add a finalizer to the pods, right?

Reply:
Stop is called when there is an eviction, not when the pods finish themselves.

The only thing that could potentially delete the finished pods is the pod gc, but the default limit is very high, so that it wouldn't affect this E2E suite.

Reply:
That makes sense.